### DISCUSSION TITLE
How do I Use async guard for 'after' by 'invoke'.

### DISCUSSION THREAD

**Discussion by elinoxp** [statelyai/xstate#3883] [Q&A] [answered]
I can use async guard for 'on' by 'invoke', but I don't really know how to use async guard for 'after' by 'invoke'. Because 'on' is executed in the order of json configuration, 'after' is executed in the order of delay time.

such as this:

```
import {  createMachine } from "xstate";   
const lightDelayMachine = createMachine(
  {
    id: "lightDelay",
    initial: "green",
    context: {
      trafficLevel: "low",
    },
    states: {
      green: {
        after: {
          LIGHT_DELAY: { target: "yellow" },
        },
      },
      yellow: {
        after: [
          {
            cond:  () => Promise.resolve(false),
            delay: 2000,
            target: "red",
          },
          {
            delay: "LIGHT_DELAY",
            target: "red",
            cond:  () => Promise.resolve(false)
          },
          {
            delay: 3000,
            target: "green",
          },
        ],
      },
      red: {
        after: {
          // 2 秒后，过渡到 green
          2000: { target: "green" },
        },
      },
    },
  },
  {
    // 此处配置的字符串延迟
    delays: {
      LIGHT_DELAY: (context, event) => {
        return context.trafficLevel === "low" ? 1000 : 3000;
      },
      YELLOW_LIGHT_DELAY: 500, // 静态值
    },
  }
);
```

**Comment by Andarist** [comment, marked as answer]
Async guards are not supported at the moment. Any async job can succeed or fail, other events might happen between the async job's start and its fulfillment, and you can have multiple **exclusive** conditions in your transition definition. All of this shows how such  a feature would complicate the design and how support for this could make it "too easy" to use async guards when they hide a lot of complexity that should be thought of during the modelling process.
